### PR TITLE
Lift key import out of `get_sas_connector`

### PR DESCRIPTION
**Pull request by daprilik**
This should fix the issue where module identities were not working with DPS-TPM provisioning.
I've also included some light refactoring, which should make the code a bit easier to grok in the future.

### CODE REVIEW

File: identity/aziot-hub-client-async/src/lib.rs [hunk 1]
```diff
@@ -275,6 +276,7 @@ impl Client {
         let body = hyper::body::to_bytes(body)
             .await
             .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+        log::debug!("IoTHub response body {:?}", body);
```
Comment (arsing) [inline identity/aziot-hub-client-async/src/lib.rs:279]: Remove this.

File: identity/aziot-cloud-client-async-common/src/lib.rs [hunk 2]
```diff
@@ -133,12 +117,10 @@ impl KeyClient for aziot_key_client_async::Client {
 
 #[async_trait::async_trait]
 impl KeyClient for aziot_tpm_client_async::Client {
+    // At the moment, we only store a single key in the TPM, so there isn't any
```
Comment (arsing) [inline identity/aziot-cloud-client-async-common/src/lib.rs:120]: Remove "At the moment". There'll only ever be one key.

File: identity/aziot-cloud-client-async-common/src/lib.rs [hunk 3]
```diff
@@ -11,18 +11,16 @@ use http_common::MaybeProxyConnector;
 pub const ENCODE_SET: &percent_encoding::AsciiSet = &http_common::PATH_SEGMENT_ENCODE_SET.add(b'=');
 
 /// `key_client` must be either an `aziot_key_client_async::Client` or an `aziot_tpm_client_async::Client`.
-pub async fn get_sas_connector(
+pub async fn get_sas_connector<K: KeyClient>(
```
Comment (arsing) [inline identity/aziot-cloud-client-async-common/src/lib.rs:14]: Put it in a `where` clause for consistency.

**Review by arsing — APPROVED** [review]
Approved with nits.

File: identity/aziot-hub-client-async/src/lib.rs (hunk 1 above)
Comment (daprilik) [reply identity/aziot-hub-client-async/src/lib.rs:279]: I added this to mirror https://github.com/Azure/iot-identity-service/blob/main/identity/aziot-dps-client-async/src/lib.rs#L309.

It proved to be reasonably useful when debugging, so I'd like to keep it in. Alternatively, if you want it gone, we should get rid of the other one as well.

Comment (arsing) [reply identity/aziot-hub-client-async/src/lib.rs:279]: @massand I vaguely remember you saying we shouldn't log cloud responses that had secrets. Is that correct?

Comment (massand) [reply identity/aziot-hub-client-async/src/lib.rs:279]: We shouldn't log any responses from IoT Hub due to credential information disclosure issues. DPS doesn't log any keys in its responses for 2 APIs. Maybe adding a comment in DPS client about the responses being clean would help avoid copying that behavior next time.

File: identity/aziot-dps-client-async/src/lib.rs [hunk 4]
```diff
@@ -161,7 +158,7 @@ impl Client {
             .await;
         };
 
-        if matches!(auth_kind, DpsAuthKind::TpmWithAuth { .. }) {
+        if matches!(auth_kind, DpsAuthKind::TpmPostChallenge { .. }) {
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:161]:
```suggestion
        if matches!(auth_kind, DpsAuthKind::TpmDpsNonce{ .. }) {
```

Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:161]:
```suggestion
        if matches!(auth_kind, DpsAuthKind::TpmDpsNonce { .. }) {
```

File: identity/aziot-dps-client-async/src/lib.rs [hunk 5]
```diff
@@ -191,7 +188,7 @@ impl Client {
     // TPM provisioning has special 2-step challenge/response flow which is
     // basically identical to the symmetric key flow, except that the TPM client
     // is used instead of the key client. To keep things DRY, a recursive call
-    // is used, passing `DpsAuthKind::TpmWithAuth` as the auth kind.
+    // is used, passing `DpsAuthKind::TpmPostChallenge` as the auth kind.
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:191]:
```suggestion
    // is used, passing `DpsAuthKind::TpmDpsNonce` as the auth kind.
```

File: identity/aziot-dps-client-async/src/lib.rs [hunk 6]
```diff
@@ -228,29 +225,33 @@ impl Client {
...
+                    false,
+                )
+                .await?;
+                let authorization_header_value = hyper::header::HeaderValue::from_str(&token)
+                    .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+                req.headers_mut()
+                    .append(hyper::header::AUTHORIZATION, authorization_header_value);
+                connector
+            }
+            DpsAuthKind::TpmPostChallenge => {
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:245]:
```suggestion
            DpsAuthKind::TpmDpsNonce => {
```

File: identity/aziot-dps-client-async/src/lib.rs [hunk 7]
```diff
@@ -331,11 +332,15 @@ impl Client {
...
-                };
+                self.tpm_client
+                    .import_auth_key(
+                        base64::decode(reg_result.authentication_key)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+                    )
+                    .await?;
 
+                // update auth method, and recurse
+                *auth_kind = DpsAuthKind::TpmPostChallenge;
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:343]:
```suggestion
                *auth_kind = DpsAuthKind::TpmDpsNonce;
```

**Review by massand — CHANGES_REQUESTED** [review]
Lets follow the general naming convention for challenge response (which is also part of the DPS TPM docs)

File: identity/aziot-dps-client-async/src/lib.rs [hunk 8]
```diff
@@ -306,6 +307,9 @@ impl Client {
         let body = hyper::body::to_bytes(body)
             .await
             .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+
+        // Unlike IoT Hub, DPS responses don't contain any potentially sensitive
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:311]: I'll eat my own words now looking at the response for TPM registration. DPS responses will contain the nonce, which is used as the signing key for registrations for a limited time period. So, let us remove the DPS response body too :)

**Review by arsing — APPROVED** [review]

File: identity/aziot-dps-client-async/src/lib.rs [hunk 9]
```diff
@@ -161,7 +158,7 @@ impl Client {
             .await;
         };
 
-        if matches!(auth_kind, DpsAuthKind::TpmWithAuth { .. }) {
+        if matches!(auth_kind, DpsAuthKind::TpmDpsNonce { .. }) {
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:161]: It seems we are recursing within `request()` itself. The nonce is imported into TPM when the `auth_kind` within `request()` method changes from `Tpm` to `TpmDpsNonce`. So, why are we reimporting the nonce into the TPM again after exiting `request()` back into `register()`?

**Review by massand — CHANGES_REQUESTED** [review]
Noticed something off about the two imports happening during TPM registration. And a DPS response logging problem I just saw

File: identity/aziot-dps-client-async/src/lib.rs (hunk 9 above)
Comment (arsing) [reply identity/aziot-dps-client-async/src/lib.rs:161]: The first response from DPS contains a challenge key. The second response from DPS contains the symmetric key registered with IoT Hub.

Comment (massand) [reply identity/aziot-dps-client-async/src/lib.rs:161]: Ah yes. Thanks.

File: identity/aziot-dps-client-async/src/lib.rs [hunk 10]
```diff
@@ -331,11 +335,15 @@ impl Client {
                 let reg_result: model::TpmRegistrationResult = serde_json::from_slice(&body)
                     .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
 
-                // update auth method
-                *auth_kind = DpsAuthKind::TpmWithAuth {
-                    auth_key: reg_result.authentication_key,
-                };
+                self.tpm_client
```
Comment (massand) [inline identity/aziot-dps-client-async/src/lib.rs:338]: Maybe a comment here would remind readers (slow readers like me) about the difference between this `import_auth_key` (which is technically importing the DPS nonce and not the primary key for IoT Hub) and the other one outside.

Comment (massand) [reply identity/aziot-dps-client-async/src/lib.rs:338]: And another suggestion for future maintainers/reviewers: to add extra emphasis on the use of an auth key, I would consider renaming `aziot_identity_common::Credentials::Tpm` to `aziot_identity_common::Credentials::TpmAuthKey`

Comment (daprilik) [reply identity/aziot-dps-client-async/src/lib.rs:338]: We already had a comment in this file (at the top of the `request` method itself), so I just moved it to a more appropriate place + tweaked the wording a bit. I just linked back to the relevant docs about how the DPS TPM flow works.

Let me know if you want to change and wording.

**Review by massand — APPROVED** [review]

**Review by arsing — APPROVED** [review]

**Pull request merged**